aws/logs_monitoring/lambda_function.py

```diff
@@ -132,50 +138,112 @@ def __exit__(self, ex_type, ex_value, traceback):
             print("DatadogConnection exit: ", ex_type, ex_value, traceback)
 
 
+class DatadogHTTPClient(object):
+
+    _POST = "POST"
+    _HEADERS = {"Content-type": "application/json"}
+
+    def __init__(self, host, apiKey, timeout=1, max_retries=1, backoff=0.5):
+        self._url = "https://{}/v1/input/{}".format(host, apiKey)
+        self._session = requests.Session()
+        self._session.headers.update(self._HEADERS)
+        self._timeout = timeout if timeout > 0 else 1
+        self._max_retries = max_retries if max_retries >= 0 else 1
+        self._backoff = backoff if backoff > 0 else 0.5
+
+    def send(self, batch, metadata=None):
+        """
+        Attemps to send a log with a linear retry strategy,
+        only retries on server and network errors.
+        """
+        body = json.dumps(batch)
+        for retry in range(1 + self._max_retries):
+            if retry > 0:
+                print("Retrying, retry: {}, max: {}".format(retry, self._max_retries))
+                time.sleep(self._backoff)
+            try:
+                resp = self._session.post(
+                    self._url, data=body, params=metadata, timeout=self._timeout
+                )
+            except Exception as e:
+                # most likely a network error
+                print("Unexpected exception: {}".format(str(e)))
+                continue
+            if resp.status_code >= 500:
+                # server error
+                print(
+                    "Server error, status {}, reason: {}".format(
+                        resp.status_code, resp.reason
+                    )
+                )
+                continue
+            elif resp.status_code >= 400:
+                # client error
+                raise Exception(
+                    "client error, status: {}, reason {}".format(
+                        resp.status_code, resp.reason
+                    )
+                )
+            else:
+                # success
+                return
+        raise Exception("max number of retries reached: {}".format(self._max_retries))
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, ex_type, ex_value, traceback):
+        self._session.close()
+        return self
+
+
+class DatadogBatcher(object):
+    def __init__(self, max_size=25):
+        self._max_size = max_size
+
+    def batch(self, logs):
+        """
+        Returns an array of batches,
+        each batch contains at most max_size logs.
+        """
+        batches = []
+        if len(logs) % self._max_size != 0:
+            nb_batchs = int(len(logs) / self._max_size) + 1
```

Comment:
always do `+1`, worst case the last batch will be empty (`a = [], a[10:20]` works fine and return an empty list)